### PR TITLE
fix(ios): fv: replace Zip framework to prevent crash on startup 🍒 

### LINKED ISSUE #8861: bug(iOS): Crash while trying to open FirstVoices App in the iPhone / iPad Simulators
### Describe the bug

I am getting crash while trying to open the installed FirstVoices application in both iPhone and iPad Simulators. 

### Reproduce the bug

1. Install FirstVoices 17.0.111-alpha build in iPhone device (?) or Simulator.
2. Click on the FirstVoices icon.
3. Crash message appear after clicking the FirstVoices application. 

I have attached the Screenshots for your reference. 

..Crash in iPhone 14 Pro Max Simulator

..Crash in iPad (10th generation) Simulator

### Expected behavior

By clicking the FirstVoices icon from the Display, the application should open without showing any error message. 

### Related issues

_No response_

### Keyman apps

- [ ] Keyman for Android
- [X] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.111-alpha

### Operating system

iOS 15.0, iOS 16.4

### Device

iPhone 14 Pro Max Simulator, iPad (10 generation) Simulator

### Target application

FirstVoices App

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### LINKED ISSUE #9819: bug(ios): FV Keyboards stable-16.0.142 app crashes on launch
Per Apple review, the stable-16.0.142 build of FV Keyboards app crashed on launch. There are no code changes from 16.0.141. Is it possible this is due to an XCode upgrade since we last built the stable release?

According to the attached reports, the Zip.framework dependency is missing:

```
"Library not loaded: @rpath\/Zip.framework\/Zip",
"Referenced from: <72FF9EF0-6AC4-37C9-AC53-394B27DF552E> \/Volumes\/VOLUME\/*\/FirstVoices.app\/Frameworks\/KeymanEngine.framework\/KeymanEngine",
"Reason: tried: 
'\/usr\/lib\/swift\/Zip.framework\/Zip' (no such file, not in dyld cache), 
'\/private\/preboot\/Cryptexes\/OS\/usr\/lib\/swift\/Zip.framework\/Zip' (no such file), 
'\/private\/var\/containers\/Bundle\/Application\/ECC845CE-BA49-4217-9DFD-1605188C26DA\/FirstVoices.app\/Frameworks\/Zip.framework\/Zip' (no such file), 
'\/private\/var\/containers\/Bundle\/Application\/ECC845CE-BA49-4217-9DFD-1605188C26DA\/FirstVoices.app\/Frameworks\/KeymanEngine.framework\/Frameworks\/Zip.framework\/Zip' (no such file), 
'\/private\/var\/containers\/Bundle\/Application\/Frameworks\/Zip.framework\/Zip' (no such file), 
'\/usr\/lib\/swift\/Zip.framework\/Zip' (no such file, not in dyld cache), 
'\/private\/preboot\/Cryptexes\/OS\/usr\/lib\/swift\/Zip.framework\/Zip' (no such file), 
'\/private\/var\/containers\/Bundle\/Application\/ECC845"
```

---

Guideline 2.1 - Performance - App Completeness

We were unable to review your app as it crashed on launch. We have attached detailed crash logs to help troubleshoot this issue.

Review device details: 

- Device type: iPad 
- OS version: iOS 17.0.3

Next Steps

To address the crash in your app, follow these steps:

1. Fully symbolicate the crash report. See [Adding Identifiable Symbol Names to a Crash Report](https://developer.apple.com/documentation/xcode/diagnosing_issues_using_crash_reports_and_device_logs/adding_identifiable_symbol_names_to_a_crash_report) for an explanation of the symbolication process.
2. Match the crash report to a common pattern. Based on the pattern, take specific actions to further investigate the crash. See [Identifying the Cause of Common Crashes](https://developer.apple.com/documentation/xcode/diagnosing_issues_using_crash_reports_and_device_logs/identifying_the_cause_of_common_crashes).
3. Test your app on a device to ensure that it now runs as expected.
4. Once you’ve addressed the crash, create and submit a new build to the App Store for review.

Resources

For additional information on crash reports, see [Diagnosing Issues Using Crash Reports and Device Logs](https://developer.apple.com/documentation/xcode/diagnosing_issues_using_crash_reports_and_device_logs).

---

* [crashlog-905474A5-F733-4C77-A738-EC2016ED4285.txt](https://github.com/keymanapp/keyman/files/13058736/crashlog-905474A5-F733-4C77-A738-EC2016ED4285.txt)
* [crashlog-DD07E783-BEA5-42B4-B789-109E00E5E802.txt](https://github.com/keymanapp/keyman/files/13058738/crashlog-DD07E783-BEA5-42B4-B789-109E00E5E802.txt)
* [crashlog-7F62B778-16EC-4387-B17C-B5249CFC8852.txt](https://github.com/keymanapp/keyman/files/13058739/crashlog-7F62B778-16EC-4387-B17C-B5249CFC8852.txt)
* [crashlog-848F31F5-3C41-4B5F-B01D-779AFD184247.txt](https://github.com/keymanapp/keyman/files/13058740/crashlog-848F31F5-3C41-4B5F-B01D-779AFD184247.txt)

### PR DESCRIPTION
**Pull request by sgschantz**
Replacing marmelroy/Zip framework with weichsel/ZIPFoundation due to problems with using the [Zip framework and Carthage](https://github.com/marmelroy/Zip/issues/70). Attempts to use Zip framework were unsuccessful in #9942 

Fixes #8861
(#9819 is specific to Stable 16.0, so that is directly fixed by #9958)

@keymanapp-test-bot skip

🍒 from #9958

### CODE REVIEW

File: ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift [hunk 1]
```diff
@@ -58,6 +58,98 @@ class KeymanPackageTests: XCTestCase {
...
+  func testKeyboardPackage_clearNonexistentDirectory_doesNothing() throws {
+    let cacheDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
+    let destinationDirectory = cacheDirectory.appendingPathComponent("doesnotexist")
+
+    do {
+      // clear directory
+      try KeymanPackage.clearDirectory(destination: destinationDirectory)
+      } catch {
+        XCTFail("error clearing the nonexistent directory \(error)")
+      }
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift:70]:
```suggestion
    } catch {
      XCTFail("error clearing the nonexistent directory \(error)")
    }
```

File: ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift [hunk 2]
```diff
@@ -58,6 +58,98 @@ class KeymanPackageTests: XCTestCase {
...
+        atPath: destinationDirectory.path,
+        withIntermediateDirectories: false,
+        attributes: nil
+      )
+
+      // clear directory
+      try KeymanPackage.clearDirectory(destination: destinationDirectory)
+      } catch {
+        XCTFail("error clearing the empty directory \(error)")
+      }
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift:89]:
```suggestion
    } catch {
      XCTFail("error clearing the empty directory \(error)")
    }
```

File: ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift [hunk 3]
```diff
@@ -58,6 +58,98 @@ class KeymanPackageTests: XCTestCase {
...
+    
+    let fileArray = try FileManager.default.contentsOfDirectory(atPath: destinationDirectory.path)
+    XCTAssert(fileArray.count == 0, "directory still contains /(fileArray.count) items")
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift:92]:
```suggestion
    XCTAssert(fileArray.count == 0, "directory still contains \(fileArray.count) items")
```

File: ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift [hunk 4]
```diff
@@ -58,6 +58,98 @@ class KeymanPackageTests: XCTestCase {
...
+      // add some files
+      FileManager.default.createFile(atPath: destinationDirectory.appendingPathComponent("fileone").path, contents: nil)
+      FileManager.default.createFile(atPath: destinationDirectory.appendingPathComponent("filetwo").path, contents: nil)
+      FileManager.default.createFile(atPath: destinationDirectory.appendingPathComponent("filethree").path, contents: nil)
+
+      // clear directory
+      try KeymanPackage.clearDirectory(destination: destinationDirectory)
+      } catch {
+        XCTFail("error clearing the empty directory \(error)")
+      }
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift:116]:
```suggestion
    } catch {
      XCTFail("error clearing the empty directory \(error)")
    }
```

File: ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift [hunk 5]
```diff
@@ -58,6 +58,98 @@ class KeymanPackageTests: XCTestCase {
...
+    
+    let fileArray = try FileManager.default.contentsOfDirectory(atPath: destinationDirectory.path)
+    XCTAssert(fileArray.count == 0, "directory still contains /(fileArray.count) items")
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift:119]:
```suggestion
    XCTAssert(fileArray.count == 0, "directory still contains \(fileArray.count) items")
```

File: ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift [hunk 6]
```diff
@@ -58,6 +58,98 @@ class KeymanPackageTests: XCTestCase {
...
+  }
+
+  func testKeyboardPackage_extractTwice_noDuplicateFileError() throws {
+    let cacheDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
+    let khmerPackageZip = cacheDirectory.appendingPathComponent("khmer_angkor.kmp.zip")
+    try ResourceFileManager.shared.copyWithOverwrite(from: TestUtils.Keyboards.khmerAngkorKMP, to: khmerPackageZip)
+
+    let destinationFolderURL = cacheDirectory.appendingPathComponent("khmer_angkor")
+
+    // Requires that the source file is already .zip, not .kmp.  It's a ZipUtils limitation.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift:129]: Is this restriction still true with the zip framework change? (If not, then that's a follow-up PR on alpha that we could pop in)

File: ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift [hunk 7]
```diff
@@ -58,6 +58,98 @@ class KeymanPackageTests: XCTestCase {
...
+              log.info("*** second unzip of \(secondKmp.id)")
+            } else {
+              XCTAssert(false, "*** second unzip failed")
+            }
+          } catch {
+            XCTFail("unzip 2 failure with error \(error)")
+          }
+       } else {
+          XCTAssert(false, "*** first unzip failed")
+      }
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift:147]:
```suggestion
        do {
          // clear directory before second extract
          try KeymanPackage.clearDirectory(destination: destinationFolderURL)
          
          if let secondKmp = try KeymanPackage.extract(fileUrl: khmerPackageZip, destination: destinationFolderURL) {
            log.info("*** second unzip of \(secondKmp.id)")
          } else {
            XCTAssert(false, "*** second unzip failed")
          }
        } catch {
          XCTFail("unzip 2 failure with error \(error)")
        }
      } else {
        XCTAssert(false, "*** first unzip failed")
      }
```

File: ios/engine/KMEI/KeymanEngineTests/TestUtils/EngineStateBundler.swift [hunk 8]
```diff
@@ -64,12 +65,19 @@ extension TestUtils {
...
-      return attachment
+      let archiveURL = bundleConstructionURL.appendingPathComponent("bundleArchive.zip")
+      do {
+        _ = try Archive(url: archiveURL, accessMode: .create)
+        log.info("archiveURL: \(archiveURL)")
+        let attachment = XCTAttachment(contentsOfFile: archiveURL)
+        attachment.lifetime = .keepAlways
+        return attachment
+      }
+      catch let error {
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/TestUtils/EngineStateBundler.swift:76]:
```suggestion
      } catch let error {
```

**Review by mcdurdin — APPROVED** [review]
LGTM, only nits.

We'll need to make sure these change shake out well in 17.0-alpha before back-porting to 16.0-stable. And then we'll need to note the breaking change in the 16.0-stable release cycle with the changed dependency. (And of course our release notes for 17.0 need to note the changed dep as well)

File: ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift (hunk 6 above)
Comment (sgschantz) [reply ios/engine/KMEI/KeymanEngineTests/KeymanPackageTests.swift:129]: I omitted '.zip' from the archive name in the unit tests, and they all passed ok -- so the new framework does not have that limitation.  I'll change in the app as a separate PR.

**Pull request merged**